pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/BaseSegmentAssignment.java

```diff
@@ -85,70 +89,41 @@ public void init(HelixManager helixManager, TableConfig tableConfig) {
       _logger.info("Initialized with replication: {} and partition column: {} for table: {}", _replication,
           _partitionColumn, _tableNameWithType);
     }
+    setSegmentAssignmentStrategyMap(instancePartitionsMap);
   }
 
   /**
-   * Returns the replication of the table.
+   * Sets Segment assignment strategy for different instance partitions and puts into a map of
+   * Map<InstancePartitionsType, SegmentAssignmentStrategy>
    */
-  protected abstract int getReplication(TableConfig tableConfig);
+  protected void setSegmentAssignmentStrategyMap(
+      Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap) {
+      _assignmentStrategyMap = SegmentAssignmentStrategyFactory.getSegmentAssignmentStrategy(_helixManager,
+          _tableConfig, instancePartitionsMap);
+  }
 
   /**
-   * Helper method to check whether the number of replica-groups matches the table replication for replica-group based
-   * instance partitions. Log a warning if they do not match and use the one inside the instance partitions. The
-   * mismatch can happen when table is not configured correctly (table replication and numReplicaGroups does not match
-   * or replication changed without reassigning instances).
+   * Returns the replication of the table.
    */
-  protected void checkReplication(InstancePartitions instancePartitions) {
-    int numReplicaGroups = instancePartitions.getNumReplicaGroups();
-    if (numReplicaGroups != _replication) {
-      _logger.warn(
-          "Number of replica-groups in instance partitions {}: {} does not match replication in table config: {} for "
-              + "table: {}, using: {}", instancePartitions.getInstancePartitionsName(), numReplicaGroups, _replication,
-          _tableNameWithType, numReplicaGroups);
-    }
-  }
+  protected abstract int getReplication(TableConfig tableConfig);
 
   /**
    * Helper method to assign instances based on the current assignment and instance partitions.
    */
   protected List<String> assignSegment(String segmentName, Map<String, Map<String, String>> currentAssignment,
```

Comment:
I think that we can remove this helper method and directly call the following line in `OfflineSegmentAssignment/RealtimeSegmentAssignment`
```
segmentAssignmentStrategy.assignSegment(segmentName,
        currentAssignment, instancePartitions, instancePartitionsType);
```